craft_archives/repo/apt_sources_manager.py

```diff
@@ -318,21 +319,58 @@ def install_package_repository_sources(
             changed = self._install_sources_apt(package_repo=package_repo)
             architectures = package_repo.architectures
             if changed and architectures:
-                _add_architecture(architectures, root=self._signed_by_root)
+                _add_architecture(
+                    architectures=architectures,
+                    root=self._signed_by_root,
+                    sources_dir=self._sources_list_d,
+                )
             return changed
 
         raise RuntimeError(f"unhandled package repository: {package_repository!r}")
 
 
-def _add_architecture(architectures: List[str], root: Path) -> None:
-    """Add package repository architecture."""
+def _add_architecture(
+    architectures: List[str],
+    root: Path,
+    sources_dir: Path,
+) -> None:
+    """Add package repository architecture.
+
+    For systems that who default to deb822 sources, `dpkg --add-architecture`
+    works for any architecture. For systems that use the traditional sources.list,
+    only compatible architectures can be added.
+
+    :param architectures: The architectures to add.
+    :param root: The root of the system to add the architectures to.
+    :param sources_dir: The directory containing the sources listings.
+    """
+    # a mapping of architectures to a list of architectures with compatible packages
+    compatible_archs = {
```

Comment:
This can probably be a constant. If you're worried about people modifying it, make it a [mapping proxy](https://docs.python.org/3/library/types.html#types.MappingProxyType).

Reply:
I moved it back to a constant.  I decided not to lock it down with a mapping proxy because there could be a valid use for modifying it and I'm ok if someone decided to do that.